python/Ganga/Utility/execute.py

```diff
@@ -98,26 +137,46 @@ def execute(command,
             return_code=None):
     """
     Execute an external command.
+    This will execute an external python command when shell=False or an external bash command when shell=True
+    Args:
+        command (str): This is the command that we want to execute in string format
+        timeout (int): This is the timeout which we want to assign to a funtion and it will be killed if it runs for longer than n seconds
+        env (dict): This is the environment to use for launching the new command
+        cwd (str): This is the cwd the command is to be executed within.
+        shell (bool): True for a bash command to be executed, False for a command to be executed within Python
+        python_setup (str): A python command to be executed beore the main command is
+        eval_includes (str): A string to be executed in the namespace of the output of the command
+        update_env (bool): Should we update the env being passed to what the env was after the command finished running
+        return_code (int): This is the returned code from the command which is executed
     """
 
     if update_env and env is None:
         raise GangaException('Cannot update the environment if None given.')
 
-    stream_command = 'cat<&0 | sh'
+    temp_pkl_file = None
+    temp_env_file = None
+
     if not shell:
+        # We want to run a pyhton command inside a small Python wrapper
         stream_command = 'python -'
-        command, pkl_read, pkl_write, envread, envwrite = python_wrapper(command, python_setup, update_env)
-    elif update_env:
-        # note the exec gets around the problem of indent and base64 gets
-        # around the \n
-        command_update, envread, envwrite = env_update_script()
-        command += ''';python -c "from __future__ import print_function;import base64;exec(base64.b64decode('%s'))"''' % base64.b64encode(command_update)
-
+        command, temp_pkl_file, temp_env_file = python_wrapper(command, python_setup, update_env)
+    else:
+        # We want to run a shell command inside a _NEW_ shell environment.
+        # i.e. What I run here I expect to behave in the same way from the command line after I exit Ganga
+        stream_command = "bash -i "
```

Comment:
Not entirely clear. Tbh If I'm going back to using a pipe method I'm going to add some debugging (and the documentation on this class has helped)